### PR TITLE
Implement "Branch protection rules" support for "publish:github" action

### PR DESCRIPTION
**Pull request by tb137**
## Hey, I just made a Pull Request!

Small change in the built-in "publish:github" action. 
Expanded action with "Branch protection rules" support capabilities.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/scaffolder-backend/src/scaffolder/actions/types.ts [hunk 1]
```diff
@@ -73,3 +73,11 @@ export type TemplateAction<Input extends JsonObject> = {
...
   handler: (ctx: ActionContext<Input>) => Promise<void>;
 };
+
+export type BypassPullRequestAllowances =
+  | {
+      users?: string[];
+      teams?: string[];
+      apps?: string[];
+    }
+  | undefined;
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/types.ts:83]: This should probably go right next to the github action, not in this general types file. Also, should not need to be `| undefined` _here_,  that would be up to the usage site to decide instead.

**Review by freben — COMMENTED** [review]
Hi, thanks! Remember to enable automatic prettier formatting in your editor, or do it manually before committing, to appease the linter.

File: plugins/scaffolder-backend/src/scaffolder/actions/types.ts (hunk 1 above)
Comment (tb137) [reply plugins/scaffolder-backend/src/scaffolder/actions/types.ts]: I use this type in few files, so not sure it's a good idea to add same type to each file where it's used... Unless you have an idea where it should go specifically?

Comment (freben) [reply plugins/scaffolder-backend/src/scaffolder/actions/types.ts]: just something like `actions/builtin/github/types.ts` or so, doesn't matter much but then it'll live with decidedly github related things

**Comment by freben** [comment]
Note that the build logs tell you to generate the API reports, since the public surfaces changed.

File: plugins/scaffolder-backend/src/scaffolder/actions/types.ts (hunk 1 above)
Comment (TomBalt) [reply plugins/scaffolder-backend/src/scaffolder/actions/types.ts]: > This should probably go right next to the github action, not in this general types file. Also, should not need to be `| undefined` _here_, that would be up to the usage site to decide instead.

I have tried to remove `| undefined`, but I get this error:
```
backstage/plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubRepoPush.ts",

Argument of type 'BypassPullRequestAllowances | undefined' is not assignable to parameter of type 'BypassPullRequestAllowances'.
  Type 'undefined' is not assignable to type 'BypassPullRequestAllowances'.
```

**Comment by TomBalt** [comment]
> Note that the build logs tell you to generate the API reports, since the public surfaces changed.

Hi @freben, 
Not sure why test is still failing as file plugins/scaffolder-backend/api-report.md was updated after running report, but it still shows a warning message and test fails :/ Would you know why?

T.

P.S. I have moved my fork to personal account, that's why user have changed.

**Comment by TomBalt** [comment]
> I also see that this now runs into DCO problems because of unsigned commits.
> 
> Having multiple identities in here will probably complicate the legal part of the signatures etc. It may be worth reconstructing the PR brand new to get around this, unless you have some good idea of how to retrofit the right signatures across the chain.
> 
> As for the build, it says `Error: The API Report for plugins/scaffolder-backend is not allowed to have warnings` - so the changes you made lead to an api report file that has warnings in it. You will have to address those warnings.

Okay... Created new repo and PR - https://github.com/backstage/backstage/pull/14320.

Current [PC](https://github.com/backstage/backstage/pull/14261) can be deleted.

**Pull request closed without merging**